services/src/main/java/org/keycloak/protocol/oidc/grants/device/endpoints/DeviceEndpoint.java

```diff
@@ -0,0 +1,488 @@
+        }
+
+        return provider.createOAuth2DeviceVerifyUserCodePage();
+    }
+
+    public Response processVerification(AuthenticationSessionModel authenticationSession, String userCode) {
+        if (userCode == null) {
+            event.error(Errors.INVALID_OAUTH2_USER_CODE);
+            return createVerificationPage(Messages.OAUTH2_DEVICE_INVALID_USER_CODE, authenticationSession);
+        }
+
+        // Format inputted user code
+        OAuth2DeviceUserCodeProvider userCodeProvider = session.getProvider(OAuth2DeviceUserCodeProvider.class);
+        String formattedUserCode = userCodeProvider.format(userCode);
+
+        // Find the token from store
+        OAuth2DeviceTokenStoreProvider store = session.getProvider(OAuth2DeviceTokenStoreProvider.class);
+        OAuth2DeviceCodeModel deviceCodeModel = store.getByUserCode(realm, formattedUserCode);
+        if (deviceCodeModel == null) {
+            event.error(Errors.INVALID_OAUTH2_USER_CODE);
+            return createVerificationPage(Messages.OAUTH2_DEVICE_INVALID_USER_CODE, authenticationSession);
+        }
+
+        int expiresIn = deviceCodeModel.getExpiration() - Time.currentTime();
+        if (expiresIn < 0) {
+            event.error(Errors.EXPIRED_OAUTH2_USER_CODE);
+            return createVerificationPage(Messages.OAUTH2_DEVICE_INVALID_USER_CODE, authenticationSession);
+        }
+
+        // Verification OK
+        authenticationSession.setClientNote(DeviceGrantType.OAUTH2_DEVICE_VERIFIED_USER_CODE, formattedUserCode);
+
+        // Event logging for the verification
+        event.client(deviceCodeModel.getClientId()).detail(Details.SCOPE, deviceCodeModel.getScope()).success();
+
+        OIDCLoginProtocol protocol = new OIDCLoginProtocol(session, realm, session.getContext().getUri(), headers, event);
+        return handleBrowserAuthenticationRequest(authenticationSession, protocol, false, true);
+    }
+
+    @Override
+    public Object getResource() {
+        return this;
+    }
+
+    @Override
+    public void close() {
+
+    }
+
+    private ClientModel authenticateClient() {
+        // https://tools.ietf.org/html/draft-ietf-oauth-device-flow-15#section-3.1
+        // The spec says "The client authentication requirements of Section 3.2.1 of [RFC6749]
+        // apply to requests on this endpoint".
+        AuthorizeClientUtil.ClientAuthResult clientAuth = AuthorizeClientUtil.authorizeClient(session, event);
+        ClientModel client = clientAuth.getClient();
+
+        if (client == null) {
+            event.error(Errors.INVALID_REQUEST);
+            throw new ErrorPageException(session, null, Response.Status.BAD_REQUEST, Messages.MISSING_PARAMETER,
+                OIDCLoginProtocol.CLIENT_ID_PARAM);
+        }
+
+        checkClient(client.getClientId(), null);
+
+        return client;
+    }
+
+    private ClientModel checkClient(String clientId, AuthenticationSessionModel authenticationSession) {
```

Comment:
There is big amount of duplicated code needed to handle client verification. Isn't it possible to re-use some basic checks (EG. call to client authentication, checking if client enabled, if OIDC login protocol, if bearer-only etc)? For example `AuthorizationEndpoint.checkClient` already uses lots of same verifications . Will it be better if DeviceEndpoint inherits from AuthorizationEndpoint and calls `super.checkClient` with possibly some more checks added specific to device flow? Like for example call to `client.isOAuth2DeviceAuthorizationGrantEnabled()` , which (if I am not missing something) is the only check specific to deviceFlow.

Reply:
Makes sense. As we have similar duplication in other places, refactoring should probably consider them too.

I don't see a problem in re-using some validations from there, but I won't push for this in this PR too.

We are trying to keep device flow-related code as much as possible in a single place. The idea is to look at the changes here to find gaps and improve how people can create extensions without having to touch different parts and duplicate code. For instance, from this PR you have that:

* We need a GrantType SPI (there are some initial discussions around it)
* We need to provide an SPI for extending the protocol API
* We need to provide ways to extend/re-use code from authorization and token endpoints

My point being, that we should think about this more carefully instead of extending what was not carefully thought for being extended.